sycl/tools/get_device_count_by_type.cpp

```diff
@@ -9,106 +9,204 @@
-        iRet = clGetDeviceIDs(platforms[i], device_type, 0, nullptr, &deviceCountPart);
-        if (iRet == CL_SUCCESS) {
-            deviceCount += deviceCountPart;
-        }
+static bool queryCUDA(cl_device_type deviceType, cl_uint &deviceCount,
+                      std::string &msg) {
+  deviceCount = 0u;
+#ifdef USE_PI_CUDA
+  const unsigned int defaultFlag = 0;
+  CUresult err = cuInit(defaultFlag);
+  if (err != CUDA_SUCCESS) {
+    msg = "CUDA initialization error";
+    return false;
+  }
+
+  const int minRuntimeVersion = 10010;
+  int runtimeVersion = 0;
+  err = cuDriverGetVersion(&runtimeVersion);
+  if (err != CUDA_SUCCESS) {
+    msg = "CUDA error querying driver version";
+    return false;
+  }
+
+  if (runtimeVersion < minRuntimeVersion) {
+    std::stringstream stream;
+    stream << "CUDA version not supported " << runtimeVersion;
+    msg = stream.str();
+    return false;
+  }
+
+  switch (deviceType) {
+  case CL_DEVICE_TYPE_DEFAULT: // Fall through.
+  case CL_DEVICE_TYPE_ALL:     // Fall through.
+  case CL_DEVICE_TYPE_GPU: {
+    int count = 0;
+    CUresult err = cuDeviceGetCount(&count);
+    if (err != CUDA_SUCCESS || count < 0) {
+      msg = "CUDA error querying device count";
+      return false;
     }
 
-    std::cout << deviceCount << ":" << backend << std::endl;
+    deviceCount = static_cast<cl_uint>(count);
+    msg = "cuda ";
+    msg += deviceTypeToString(deviceType);
+    return true;
+  } break;
+  default:
+    msg = "CUDA unsupported device type ";
+    msg += deviceTypeToString(deviceType);
+    return false;
+  }
+#else
+  msg = "CUDA not supported";
+  deviceCount = 0u;
+
+  return false;
+#endif
+}
 
-    return 0;
+int main(int argc, char *argv[]) {
+  if (argc < 3) {
+    std::cout << "0:Please set a device type and backend to find" << std::endl
+              << help << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  std::string type = argv[1];
+  std::string backend{argv[2]};
+
+  cl_device_type deviceType = CL_DEVICE_TYPE_DEFAULT;
+  if (type == "cpu") {
+    deviceType = CL_DEVICE_TYPE_CPU;
+  } else if (type == "gpu") {
+    deviceType = CL_DEVICE_TYPE_GPU;
+  } else if (type == "accelerator") {
+    deviceType = CL_DEVICE_TYPE_ACCELERATOR;
+  } else if (type == "default") {
+    deviceType = CL_DEVICE_TYPE_DEFAULT;
+  } else if (type == "all") {
+    deviceType = CL_DEVICE_TYPE_ALL;
+  } else {
+    std::cout << "0:Incorrect device type " << type << "\n"
+              << help << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  std::string msg;
+  cl_uint deviceCount = 0;
+
+  bool querySuccess = false;
+
+  if (backend == "opencl" || backend == "OpenCL" || backend == "OPENCL" ||
+      backend == "PI_OPENCL") {
+    querySuccess = queryOpenCL(deviceType, deviceCount, msg);
+  } else if (backend == "cuda" || backend == "CUDA" || backend == "PI_CUDA") {
+    querySuccess = queryCUDA(deviceType, deviceCount, msg);
+  } else {
+    std::stringstream stream;
```

Comment:
Why stream is required here? Can you just append string?

Reply:
Good point, will simplify.